fix(css) Update `ray()` and `offset-position` text and examples for default `offset-position: normal`

dipikabh opened the pull request:
<!-- 🙌 Thanks for contributing to MDN Web Docs. Adding details below will help us to merge your PR faster. -->

### Description

The initial value of [`offset-position`](https://developer.mozilla.org/en-US/docs/Web/CSS/offset-position) property is now `normal`. It was earlier `auto`.

Change in the default value of `offset-position` can be tracked here:

- Spec: https://drafts.fxtf.org/motion/#offset-position-property
- Spec update: https://github.com/w3c/fxtf-drafts/commit/fc81ee1361077c9f14cc73c3f99346f704762d4d
- Change implemented in Firefox:
       - https://bugzilla.mozilla.org/show_bug.cgi?id=1835434
       - https://bugzilla.mozilla.org/show_bug.cgi?id=1559232

### Motivation

The examples and text on `offset-position` and [`ray()`](https://developer.mozilla.org/en-US/docs/Web/CSS/ray) pages need to be adjusted for the default value of `offset-position` as `normal`.

| Current [ray() example](https://developer.mozilla.org/en-US/docs/Web/CSS/ray#defining_the_angle_and_starting_position_for_a_ray) | After fixes in this PR (Fx122) |
|---|---|
| <kbd>![Screenshot 2024-01-22 at 3 50 10 PM](https://github.com/mdn/content/assets/23019147/2bc4112d-0a26-4cfa-b6a3-dcda2a0a7d4f)</kbd> | <kbd>![Screenshot 2024-01-22 at 3 51 29 PM](https://github.com/mdn/content/assets/23019147/6cc829b0-e0ab-427d-b3b3-a957373813dc)</kbd> |

### Additional details

- Examples tested in ****Firefox 122 beta** (releasing on Jan 23)** and Chrome 120
- Codepen: https://codepen.io/dipikabh/pen/ZEPKVGZ?editors=1100

### Related issues and pull requests

PR to update the "Initial value" of `offset-position`: https://github.com/mdn/data/pull/709

dipikabh committed:
updates offset-position default value

dipikabh committed:
updates offset-position example descriptions

dipikabh committed:
updates ray examples and descriptions

dipikabh commented:
Hi @estelle, could you take a look at this PR when you get the chance? Thank you!

estelle commented:
Sorry for the delay. I am in the middle of two large PR reviews, and will get to this after I finish those.

pepelsbey reviewed (commented):
It looks good to me, thanks! But I also noticed something in the text that came before your PR: the value `50%, 50%` is mentioned multiple times in different ways:

- `(50%, 50%)`
- (`50%, 50%`)
- `50%, 50%`

I believe it’s the default `normal` value for the offset position that equals `50% 50%` without a comma or parenthesis. I wonder if we should normalize it to the same format that `offset-position` takes as a value.

dipikabh commented:
Hi @estelle, no worries, I didn't mean to rush you. I also didn't want this PR to go stale or have issues being filed because the current examples look wonky.

Thanks for the review, @pepelsbey!

Good observation about the `normal` value.

> `50% 50%` without a comma or parenthesis 

I believe I've been doing this but _with_ the parenthesis. I'll take a look and check the instances that'll need fixing.

estelle reviewed (commented):
some of my comments are about stuff that is already live

dipikabh commented on files/en-us/web/css/offset-position/index.md:
```diff
@@ -7,7 +7,7 @@ browser-compat: css.properties.offset-position
 {{CSSRef}}
 
-The **`offset-position`** [CSS](/en-US/docs/Web/CSS) property defines the initial position of an element along a path. This property is typically used in combination with the {{cssxref("offset-path")}} property to create a motion effect. The value of `offset-position` determines where the element gets placed initially for moving along an `offset-path` if the offset-path function does not specify  ...
+The **`offset-position`** [CSS](/en-US/docs/Web/CSS) property defines the initial position of an element along a path. This property is typically used in combination with the {{cssxref("offset-path")}} property to create a motion effect. The value of `offset-position` determines where the element gets placed initially for moving along an offset path if the `offset-path` functions such as {{cssxre ...
```
I've changed it to singular and `path` function, also removed commas around "such as" in the below version.

> The value of `offset-position` determines where the element gets placed initially for moving along an offset path if an `offset-path` function such as {{cssxref("path", "path()")}} does not specify its own starting position.

By saying "functions", I was alluding to the fact that `offset-path` can takes functions like `path()`, `ray()`, `circle()`, `polygon()`, etc.

dipikabh commented on files/en-us/web/css/offset-position/index.md:
```diff
@@ -234,7 +234,7 @@ pre {
 {{EmbedLiveSample('Comparing various offset starting positions', '100%', 930)}}
 
-Notice that when `offset-position` is `normal`, the starting position of the ray is (`50%, 50%`) of the containing block. Also notice the difference between offset starting positions `auto` and `left top`. The value `auto` places the element such that its anchor point is at the top-left corner of the element box itself, whereas the value `left top` places the element such that the anchor point is ...
+In box `0`, the absence of the `offset-path` property means setting `offset-position` to either `normal` or `auto` has no effect. Notice that when `offset-position` is `normal`, the ray starts at (`50%, 50%`) of the containing block. This is the default starting position of an offset path and is used when no `offset-position` is specified. Also notice the difference between offset starting positi ...
```
To remove the confusion about the parenthesis, as also pointed by Vadim in his [comment](https://github.com/mdn/content/pull/31850#pullrequestreview-1867561105), I've fixed this as:

> When `offset-position` is `normal`, the ray starts at the center of the containing block (i.e., `50% 50%`).

dipikabh commented on files/en-us/web/css/ray/index.md:
```diff
@@ -53,13 +53,13 @@ The parameters can be specified in any order.
 
 - `at <position>`
-  - : Specifies the point where the ray begins and where the element is placed in its containing block. This is an optional parameter. If omitted, the value used is the `offset-position` value of the element. If omitted and the element doesn't have an `offset-position` value, the value used for ray's starting position is `offset-position: auto`, which places the element at the `top left` corner o ...
+  - : Specifies the point where the ray begins and where the element is placed in its containing block. This is an optional parameter. If omitted, the value used is the `offset-position` value of the element. If omitted and the element doesn't have an `offset-position` value, the value used for ray's starting position is `offset-position: normal`, which places the element at (`50%`, `50%`) of the ...
```
Removed the comma and parentheses

dipikabh commented on files/en-us/web/css/ray/index.md:
```diff
@@ -53,13 +53,13 @@ The parameters can be specified in any order.
 The `ray()` function allows the positioning of an element along a path using polar coordinates, rather than using the standard rectangular coordinates used by the {{cssxref("translate","translate()")}} function or using animation to move an element along a defined path. The `ray()` function helps to create 2D spatial transitions.
 
-The element is initially positioned by moving the element's [`offset-anchor`](/en-US/docs/Web/CSS/offset-anchor) point to the element's offset starting position. The default offset starting position of a ray is the default `offset-position` value: `auto`. If `offset-position` is explicitly specified as `auto` or omitted and allowed to default, the offset starting position is the `top left` corner ...
+The element is initially positioned by moving the element's [`offset-anchor`](/en-US/docs/Web/CSS/offset-anchor) point to the element's offset starting position. The default offset starting position of a ray is the default {{cssxref("offset-position")}} value: `normal`. If `offset-position` is explicitly specified as `normal` or omitted and allowed to default, the offset starting position is `50% ...
```
Fixed this as:

> If `offset-position` is explicitly specified as `normal` or omitted and allowed to default, the offset starting position is the `center` (or `50% 50%`) of the element's containing block.

dipikabh commented on files/en-us/web/css/ray/index.md:
```diff
@@ -114,101 +119,99 @@ pre {
+<pre>
+  offset-position: 0 0;
+  offset-path: ray(0deg);
+  /* Starting position is top-left corner */
```
My original intent was to place all comments at the end. But you're right, we can remove this one

dipikabh commented on files/en-us/web/css/ray/index.md:
```diff
@@ -53,13 +53,13 @@ The parameters can be specified in any order.
   - : Reduces the length of the line segment so that the element stays within the containing block even at `offset-distance: 100%`. Specifically, the segment's length is reduced by half the width or half the height of the element's border box, whichever is greater, but never going less than zero. It is an optional parameter.
```
Good catches, thank you!
- Removed line 15: `/* offset-path: ray(parameters) */`
- The note is no longer true, `<size>` is not required. Checked in the latest FF, Chrome, and Safari. I've removed the note as well as `closest-side` from two boxes in the example that was added primarily for cross-browser demo.

dipikabh commented on files/en-us/web/css/ray/index.md:
```diff
@@ -53,13 +53,13 @@ The parameters can be specified in any order.
 
 - `at <position>`
-  - : Specifies the point where the ray begins and where the element is placed in its containing block. This is an optional parameter. If omitted, the value used is the `offset-position` value of the element. If omitted and the element doesn't have an `offset-position` value, the value used for ray's starting position is `offset-position: auto`, which places the element at the `top left` corner o ...
+  - : Specifies the point where the ray begins and where the element is placed in its containing block. This is an optional parameter. If omitted, the value used is the `offset-position` value of the element. If omitted and the element doesn't have an `offset-position` value, the value used for ray's starting position is `offset-position: normal`, which places the element at (`50%`, `50%`) of the ...
```
Great addition about `at`, thank you! Fixed this as:

> If included, the `<position>` value must be preceded by the `at` keyword. If omitted, the value used is the `offset-position` value of the element. If omitted and the element doesn't have an `offset-position` value, the value used for the ray's starting position is `offset-position: normal`, which places the element at the center (or `50% 50%`) of the containing block.

dipikabh commented on files/en-us/web/css/ray/index.md:
```diff
@@ -53,13 +53,13 @@ The parameters can be specified in any order.
   - : Reduces the length of the line segment so that the element stays within the containing block even at `offset-distance: 100%`. Specifically, the segment's length is reduced by half the width or half the height of the element's border box, whichever is greater, but never going less than zero. It is an optional parameter.
```
It does

dipikabh commented on files/en-us/web/css/ray/index.md:
```diff
@@ -53,13 +53,13 @@ The parameters can be specified in any order.
 
 ## Description
 
 The `ray()` function allows the positioning of an element along a path using polar coordinates, rather than using the standard rectangular coordinates used by the {{cssxref("translate","translate()")}} function or using animation to move an element along a defined path. The `ray()` function helps to create 2D spatial transitions.
```
Good one. Rewritten and explained as follows:

> The `ray()` function positions an element along a path by specifying points in a two-dimensional space through an angle and a distance from a reference point (polar coordinates). This feature makes the `ray()` function useful for creating 2D spatial transitions. For comparison, this approach differs from the method of specifying a point by its horizontal and vertical distances from a fixed origin (rectangular coordinates), which is used by the {{cssxref("translate","translate()")}} function and from moving an element along a defined path through animation.

dipikabh commented on files/en-us/web/css/ray/index.md:
```diff
@@ -114,101 +119,99 @@ pre {
 .box3 {
-  offset-anchor: 0 0;
-  offset-path: ray(150deg closest-side);
+  offset-rotate: 0deg;
```
There is a reference to the offset properties `offset-position` and `offset-path` in the intro and then to `offset-anchor` in "Description". `offset-rotate` is mentioned _after_ the example where we're explaining the "Result". 
Did you have anything else in mind?

estelle replied:
Oops. I didn't include the "rotate" half of the property.
I meant "add a sentence about `offset-rotate` somewhere in this doc before we get to this example"
It would be helpful if we can figure out a way to add `offset-rotate` into the content before we get to this point because it is not a well-known CSS property, so when someone doesn't know the property and reads the example, it isn't obvious what it does. (unlike, say, `color: red;` which doesn't need to be explained).

I will add a suggestion in the description, which i hope you will edit and make better.

dipikabh replied:
Thanks :)

dipikabh commented on files/en-us/web/css/ray/index.md:
```diff
@@ -114,101 +119,99 @@ pre {
+</pre>
 <div class="container">
   <div class="box">0</div>
   <div class="box box2">2</div>
```
Nope. I've observed the same problem (in Firefox). It was there when I last worked on this page and again forgot to ask the engineer.
In Firefox, the numbers appear when I select them in the HTML in DevTools.
In Chrome, the entire "Result" frame takes a while to load. Same in `offset-position` page example. But at least the numbers appear in Chrome. IIRC, the numbers were not displaying even in Chrome last time.

estelle replied:
I was trying to debus as well. They weren't showing up in Edge or FF for me (the only ones I tested), but when i selected them to try to debug, they showed up. So I couldn't easily trouble shoot.

dipikabh commented on files/en-us/web/css/ray/index.md:
```diff
@@ -114,101 +119,99 @@ pre {
 .box2 {
-  offset-anchor: 0 0;
-  offset-path: ray(0deg closest-side);
+  offset-path: ray(closest-side 150deg);
```
`closest-side` here was added to align with the "Note" on the page. I've removed `closest-side` as per [my previous comment](https://github.com/mdn/content/pull/31850#discussion_r1483411646).

dipikabh committed:
fixes review comments

dipikabh commented on files/en-us/web/css/ray/index.md:
```diff
@@ -114,101 +119,99 @@ pre {
 
-{{EmbedLiveSample('Giving an angle to the ray', '100%', 1000)}}
+{{EmbedLiveSample('Giving an angle to the ray', '100%', 1100)}}
+
```
Nice! I've integrated this text with the paragraph before the "Result" frame so that the prose describing the example is all together.

> In this example, various `offset-path: ray()` values are applied to the boxes numbered `1` to `5`. The "containing block" of each box is depicted with a dashed border. A faded box in the upper left corner shows each box's default position without any `offset-position` or `offset-path` applied, allowing for a side-by-side comparison. The top of each box is highlighted with a `solid` border to illustrate variations in ray starting points and orientations. After positioning at the ray's starting point, a box aligns with the direction of the specified ray angle. If {{cssxref("offset-position")}} is not specified, the default offset starting position of the ray is the center (or `50% 50%`) of the box's containing block.

dipikabh commented:
Thanks @estelle for the excellent review! I've fixed most for your suggestions and noted where my fix is slightly different from the suggestion.

@pepelsbey indeed there were instances of `50% 50%` with commas. I've removed the parentheses and commas (Estelle also pointed the specific instances). Thanks for spotting!

estelle commented on files/en-us/web/css/ray/index.md:
```diff
@@ -46,20 +44,18 @@ The parameters can be specified in any order.
+The `ray()` function positions an element along a path by specifying points in a two-dimensional space through an angle and a distance from a reference point (polar coordinates). This feature makes the `ray()` function useful for creating 2D spatial transitions. For comparison, this approach differs from the method of specifying a point by its horizontal and vertical distances from a fixed origin ...
 
-The element is initially positioned by moving the element's [`offset-anchor`](/en-US/docs/Web/CSS/offset-anchor) point to the element's offset starting position. The default offset starting position of a ray is the default `offset-position` value: `auto`. If `offset-position` is explicitly specified as `auto` or omitted and allowed to default, the offset starting position is the `top left` corner ...
+The element is initially positioned by moving the element's [`offset-anchor`](/en-US/docs/Web/CSS/offset-anchor) point to the element's offset starting position. The default offset starting position of a ray is the default {{cssxref("offset-position")}} value: `normal`. If `offset-position` is explicitly specified as `normal` or omitted and allowed to default, the offset starting position is the  ...
```
```suggestion
The element is initially positioned by moving the element's [`offset-anchor`](/en-US/docs/Web/CSS/offset-anchor) point to the element's offset starting position. The default offset starting position of a ray is the default {{cssxref("offset-position")}} value: `normal`. If `offset-position` is explicitly specified as `normal` or omitted and allowed to default, the offset starting position is the `center` (or `50% 50%`) of the element's containing block. With `offset-position: auto`, the starting position of the ray is the `top left` corner (or `0 0`) of the element's box.

When the `ray()` function is applied as the `offset-path` value on an element, by default, the element's inline axis rotates to the angle specified by the ray's angle. The `ray()` angle of `0deg` lies on the y-axis pointing up, with positive angles increasing in the clockwise direction. This means that when `0deg` is set on an element, the inline-axis will be vertical. Use the {{cssxref("offset-rotate"}} to override this rotation.
```

estelle replied:
this is my attempt

dipikabh replied:
This is great feedback - adding info about `offset-rotate`! It is an important piece and was missing an explanation. In my iteration to incorporate your suggestion, I actually ended by editing a larger chunk. This nicely ties in with the explanation added for polar coordinates yesterday. All the pieces seem to be coming together now!

What do you think about this (I've copied the entire updated "Description" here to provide context and show how the content flows; `offset-rotate` is in the second bullet):

> The `ray()` function positions an element along a path by specifying points in a two-dimensional space through an angle and a distance from a reference point (polar coordinates). This feature makes the `ray()` function useful for creating 2D spatial transitions. For comparison, this approach differs from the method of specifying a point by its horizontal and vertical distances from a fixed origin (rectangular coordinates), which is used by the {{cssxref("translate","translate()")}} function, and from moving an element along a defined path through animation.
> 
> As `ray()` works in 2D space, it's important to consider both the initial position and orientation of the element. When the `ray()` function is applied as the `offset-path` value on an element, here's how you can control these aspects:
> 
> - The element is initially positioned by moving the element's [`offset-anchor`](/en-US/docs/Web/CSS/offset-anchor) point to the element's offset starting position. By default, the ray's starting position is determined by the {{cssxref("offset-position")}} value. If `offset-position` is explicitly specified as `normal` (or omitted and allowed to default to `normal`), the element is positioned at the `center` (or `50% 50%`) of its containing block. Specifying `offset-position: auto` sets the starting position at the element's `top left` corner (or `0 0` of the element).
> 
> - The element is initially rotated such that its [inline axis](/en-US/docs/Web/CSS/CSS_grid_layout/Box_alignment_in_grid_layout#the_two_axes_of_a_grid_layout) — its direction of text flow — aligns with the angle specified by `ray()`. For example, with the `ray()` angle of `0deg`, which lies on the y-axis pointing up, the element's inline axis is rotated to be vertical to match the ray's angle. The element maintains this rotation throughout its path. To customize this behavior, use the {{cssxref("offset-rotate")}} property, which allows you to specify a different rotation angle or direction for the element, enabling more precise control over its appearance as it follows the path.
> 
> To define the spatial context for these operations, the `<coord-box>` value of the [`offset-path`](/en-US/docs/Web/CSS/offset-path) property provides the reference box for the ray path, defining where and how these movements and orientations are calculated.

estelle replied:
Really good. Just a few suggestions which I'll post as suggestions.

dipikabh committed:
fixes review feedback

estelle reviewed (commented):
Just a few suggestions, and hopefully a fix for the missing numbers

dipikabh commented on files/en-us/web/css/ray/index.md:
```diff
@@ -46,22 +44,23 @@ The parameters can be specified in any order.
+- The element is initially positioned by moving the element's [`offset-anchor`](/en-US/docs/Web/CSS/offset-anchor) point to the element's offset starting position. By default, the ray's starting position is determined by the {{cssxref("offset-position")}} value. If `offset-position` is explicitly specified as `normal` (or omitted and allowed to default to `normal`), the element is positioned at t ...
+- The element is initially rotated such that its [inline axis](/en-US/docs/Web/CSS/CSS_grid_layout/Box_alignment_in_grid_layout#the_two_axes_of_a_grid_layout) — its direction of text flow — aligns with the angle specified by `ray()`. For example, with the `ray()` angle of `0deg`, which lies on the y-axis pointing up, the element's inline axis is rotated to be vertical to match the ray's angle. Th ...
+
+To define the spatial context for these operations, the `<coord-box>` value of the [`offset-path`](/en-US/docs/Web/CSS/offset-path) property provides the reference box for the ray path, defining where and how these movements and orientations are calculated.
```
It originally read: "The <coord-box> value of the [offset-path](https://developer.mozilla.org/en-US/docs/Web/CSS/offset-path) property provides the reference box for the ray path."

Probably better to remove it from here, it makes more sense on the `offset-path` page. It's already included there and I must have added it here as well because of its connection to `ray()`.

dipikabh committed:
fixes feedback and example codes

dipikabh commented on files/en-us/web/css/ray/index.md:
```diff
@@ -71,17 +70,22 @@ The `<coord-box>` value of the [`offset-path`](/en-US/docs/Web/CSS/offset-path)
+}
+
 .container {
   position: relative;
```
It does cure the numbers glitching!! Thank you 🙌 

I applied the same logic to the `offset-page` example code because it had the same issue. The numbers started appearing here as well 🎉 

However, in the `offset-page` example --> `box2` is not getting positioned at its correct `auto` position. `box2` and `box3` seem to have the same position.
See [preview](https://pr31850.content.dev.mdn.mozit.cloud/en-US/docs/Web/CSS/offset-position#result_2) and [live](https://developer.mozilla.org/en-US/docs/Web/CSS/offset-position#result_2): The live page shows the correct `auto` position. 

I had used `top: 20px; left: 20px;` in `.box` to specifically show the difference between `auto` and `left top`. But with or without it now, `box2` and `box3` seem to have the same position.

This will need some more of your CSS magic 💟.

dipikabh committed:
reverts changes in offset-position

dipikabh commented:
> I need some more of your help with fixing offset-position example. 🙏

Hi @estelle, I've reverted the changes in the `offset-position` file to help move this PR forward (I'll tinker with it separately). I hope all your comments are addressed and we can land the updates. Thanks!

pepelsbey reviewed (approved):
LGTM, thanks!

dipikabh commented:
Thanks, @pepelsbey!
